[Ryan Tan] iP

Ryan-Tan00 opened the pull request:
# Candice

> “Your mind is for having ideas, not holding them.” – David Allen ([source](https://dansilvestre.com/productivity-quotes))

DukePro frees your mind of having to remember things you need to do. It's,

- text-based
- easy to learn
- ~~FAST~~ *SUPER* FAST to use

All you need to do is,

1. download it from [here](../../).
2. double-click it.
3. add your tasks.
4. let it manage your tasks for you 😉

And it is **FREE**!

Features:

- [x] Managing tasks
- [ ] Managing deadlines (coming soon)
- [ ] Reminders (coming soon)

##

If you Java programmer, you can use it to practice Java too. Here's the `main` method:

```java
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}
```

Commit:
Add Gradle support

Commit:
added the skeletal version of Duke that echos commands

Commit:
added the ability to store inputs into a list and display the elements in the list

Commit:
added the ability to mark tasks as done or not done

Commit:
renamed duke to Candice and added functionality for ToDo, Deadline and Event

Commit:
very slight edit to the response from candice

Commit:
added exceptions and exceptions handling

Commit:
added the delete functionality

Commit:
changed the java file name back to Duke and added test cases + their expected result

Commit:
redid the logic flow to allow for the Action class to parse the input and create Action instances that can be resolved. Also added some new test cases

Commit:
editing the exceptions slightly and adding a test case

Commit:
Add the ability to save tasks locally

Uses a text file to store the task list created during usage of the
program. Allows for task list to be reused after closing the program

HowSuen commented on src/main/java/Action.java:
```diff
@@ -0,0 +1,160 @@
+
+        String action = actionSplit[0];
+
+        switch (action) {
```
For `switch` statements, there should be no indentation for `case` and `default` clauses

HowSuen commented on src/main/java/Action.java:
```diff
@@ -0,0 +1,160 @@
+        }
+    }
+
+    public abstract void resolve(ArrayList<Task> taskList);
```
Since your sub classes are overriding this `resolve` method, perhaps it would be better to place this abstract method at the top so we can expect to see this method in the sub classes?

HowSuen commented on src/main/java/Action.java:
```diff
@@ -0,0 +1,160 @@
+        }
+    }
+
+    public static Action actionParser(String input) throws UnknownActionException, EmptyTaskNameException,
```
I like that you have this function to handle the commands given by the user. :) 
Maybe you can consider splitting the actions into separate functions and use abstraction to call them? Perhaps it would make the code for this function shorter and simpler to follow!

HowSuen commented on src/main/java/TimedTask.java:
```diff
@@ -0,0 +1,30 @@
+        }
+
+        @Override
+        public String status() {
```
Would it be better if you used verbs to name methods like these, eg. `showStatus()` instead of `status()`?

HowSuen reviewed (commented):
Overall, I like the flow of logic you used and it was easy to read and understand.
You may consider creating separate class files for your subclasses in `Action` and `Task` and package them under `action` and `task` packages for instance, as there is an increment A-Packages where you organise your classes into packages. It also makes your code more compact and less verbose. I also like the personality of your Duke bot!

Commit:
Use LocalDate and LocalTime instead of Strings for TimedTasks

Deadline and Event tasks now use LocalDate which is compulsory and
LocalTime which is optional (set to null by default) instead of a
String to encapsulate the deadline timing

Commit:
Merge branch 'branch-Level-8'

# Conflicts:
#	src/main/java/Action.java
#	src/main/java/TimedTask.java

Commit:
Add minor adjustments to optimize code

removed unneeded import statements and methods

Commit:
Change logic flow of code to make it more object oriented

Renamed Action to Command, created a CommandType Enum class, created
Storage, Paraser and Ui classes for a more object oriented code.
Still need to touch up on object oriented approach of the Duke class.

Commit:
Add packaging

Added packages to the files (candice, candice.command, candice.exception
and candice.task)

Commit:
Merge branch 'add-gradle-support'

Commit:
Add JUnit testing cases

Added test classes for every main class with public methods and added
test cases for the Task and TimedTask classes

Commit:
Add some java documentation

Added javadocs for Parser and CommandType

Commit:
Add javadocs to Command

Added java documentation for all methods and variables in the Command
class.

Commit:
Add Javadocs

Added documentation for all classes and non-abstract methods.

Commit:
Add javadocs for Ui

Added java documentation for the methods in the Ui class.

Commit:
Fix small bugs in the code

Changed the storage description and test cases of the event class as
it was falsely returning and checking the event class as a deadline
class.

Gavzzz commented on src/main/java/TimedTask.java:
```diff
@@ -1,43 +1,93 @@
         }
-
     }
 }
```
Overall code looks good and clean and should have no problem completing the IP

Commit:
Add small changes and changed code to adhere to coding standard

Removed try catch block that was redundant.
Changed boolean name from finished to isFinished.

Commit:
Add find as a command

Added the ability to find tasks.

Commit:
Merge branch 'branch-A-JavaDoc'

Commit:
Merge branch 'branch-A-CodingStandard'

# Conflicts:
#	src/main/java/candice/task/Task.java

Commit:
Merge branch 'branch-Level-9'

# Conflicts:
#	src/main/java/candice/command/Command.java
#	src/main/java/candice/exception/EmptyCommandDescriptionException.java
#	src/main/java/candice/task/TaskList.java

Ryan-Tan00 committed:
Update README.md

Ryan-Tan00 committed:
Update README.md

Ryan-Tan00 committed:
Update README.md

Ryan-Tan00 committed:
Update README.md

Commit:
Add javafx and small changes to logic flow

Changed the logic flow of storage. Creating an instance of Storage now
creates the storage file at the given path if it does not exist.

Changed the logic flow of NonTimedCommand

Commit:
Update README.md

Commit:
Merge branch 'master' of https://github.com/Ryan-Tan00/ip

Commit:
Add GUI for Candice

Commit:
Add javadocs

Commit:
Add assertions

Parts of code rely on certain arguments being exclusively one or two
values.

The arguments being inputted could potentially be outside the range of
intended arguments.

Use assertions to ensure that only the intended arguments are inputted.

Let's add assertions to ensure only intended arguments are the ones
that were inputted to reduce the number of bugs.

Commit:
Remove code duplication by adding to parent class

The getStorageDescription method duplicates code 3 times in the ToDo,
Deadline and Event class.

Deadline and Event classes both have a LocalTime variable in it.

Declaring a LocalTime variable in the TimedTask class as well as
implementing the getStorageDescription method in both Task and TimedTask
can reduce the amount of code duplication.

Let's create a LocalTime variable in the TimedTask class as well as
implement a getStorageDescription method in both the Task and TimedTask
classes.

This allows us to initialize less variables in Deadline and Event as
well as use the implementation of getStorageDescription in the parent
classes to implement the getStorageDescription method in ToDo, Deadline
and Event.

Commit:
Remove extra spaces

Commit:
Fix minor bug (extra ;)

Ryan-Tan00 committed:
Merge pull request #2 from Ryan-Tan00/branch-A-Assertions

Add assertions

Ryan-Tan00 committed:
Merge pull request #3 from Ryan-Tan00/branch-A-CodeQuality

Improve code quality

Commit:
Add undo feature

No undo feature to undo unwanted commands.

Having an undo feature allows users to undo an unwanted command.

Add an undo feature to allow users to undo the previous command that
edited the task list. If undo is inputted after an undo, the second undo
undoes the first undo (i.e. inputting "mark 1", "undo" and "undo" will
mark the 1st task as finished.).

Undo allows the latest command to be undone and if undo was the
unwanted command, undo can be inputted again to undo the undo command.

Ryan-Tan00 committed:
Update README.md

Ryan-Tan00 committed:
Update README.md

Ryan-Tan00 committed:
Update README.md

Ryan-Tan00 committed:
Update README.md

Ryan-Tan00 committed:
Update README.md

Ryan-Tan00 committed:
Update README.md

Ryan-Tan00 committed:
Update README.md

Ryan-Tan00 committed:
Update README.md

Ryan-Tan00 committed:
Update README.md

Commit:
Fix minor bugs
